File: src/main/java/cz/vhromada/catalog/gui/AboutDialog.java
package cz.vhromada.catalog.gui;

import java.awt.Font;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.GroupLayout;
import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.SwingConstants;
import javax.swing.WindowConstants;

import cz.vhromada.catalog.gui.commons.CatalogSwingConstants;
import cz.vhromada.catalog.gui.commons.Picture;

/**
 * A class represents dialog about.
 *
 * @author Vladimir Hromada
 */
public class AboutDialog extends JDialog {

    /**
     * SerialVersionUID
     */
    private static final long serialVersionUID = 1L;

    /**
     * Horizontal label size
     */
    private static final int HORIZONTAL_LABEL_SIZE = 200;

    /**
     * Horizontal button size
     */
    private static final int HORIZONTAL_BUTTON_SIZE = 76;

    /**
     * Horizontal button gap size
     */
    private static final int HORIZONTAL_BUTTON_GAP_SIZE = 62;

    /**
     * Horizontal gap size
     */
    private static final int HORIZONTAL_GAP_SIZE = 20;

    /**
     * Vertical gap size
     */
    private static final int VERTICAL_GAP_SIZE = 20;

    /**
     * Vertical long gap size
     */
    private static final int VERTICAL_LONG_GAP_SIZE = 40;

    /**
     * Font size
     */
    private static final int FONT_SIZE = 12;

    /**
     * Label for name
     */
    private JLabel nameLabel = new JLabel("Catalog", Picture.CATALOG.getIcon(), SwingConstants.LEADING);

    /**
     * Label for version
     */
    private JLabel versionLabel = new JLabel("Version: 1.0.0");

    /**
     * Label for author
     */
    private JLabel authorLabel = new JLabel("Author: Vladimír Hromada");

    /**
     * Label for copyrights
     */
    private JLabel rightsLabel = new JLabel("All rights reserved.");

    /**
     * Button OK
     */
    private JButton okButton = new JButton("OK", Picture.OK.getIcon());

    /**
     * Creates a new instance of AboutDialog.
     */
    public AboutDialog() {
        super(new JFrame(), "About", true);

        setIconImage(Picture.ABOUT.getIcon().getImage());
        setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
        setResizable(false);

        final Font font = new Font("Tahoma", Font.BOLD, FONT_SIZE);

        initLabels(font, nameLabel, versionLabel, authorLabel, rightsLabel);

        okButton.setFont(font);
        okButton.addActionListener(new ActionListener() {

            @Override
            public void actionPerformed(final ActionEvent e) {
                setVisible(false);
                dispose();
            }

        });

        final GroupLayout layout = new GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(createHorizontalLayout(layout));
        layout.setVerticalGroup(createVerticalLayout(layout));

        okButton.requestFocusInWindow();
        pack();
        setLocationRelativeTo(getRootPane());
    }

    /**
     * Initializes labels.
     *
     * @param font   font
     * @param labels labels
     */
    private static void initLabels(final Font font, final JLabel... labels) {
        for (final JLabel label : labels) {
            label.setFocusable(false);
            label.setFont(font);
        }
    }

    /**
     * Returns horizontal layout of components.
     *
     * @param layout layout
     * @return horizontal layout of components
     */
    private GroupLayout.SequentialGroup createHorizontalLayout(final GroupLayout layout) {
        final GroupLayout.ParallelGroup labels = layout.createParallelGroup()
                .addComponent(nameLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE)
                .addComponent(versionLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE)
                .addComponent(authorLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE)
                .addComponent(rightsLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE);

        final GroupLayout.SequentialGroup button = layout.createSequentialGroup()
                .addGap(HORIZONTAL_BUTTON_GAP_SIZE)
                .addComponent(okButton, HORIZONTAL_BUTTON_SIZE, HORIZONTAL_BUTTON_SIZE, HORIZONTAL_BUTTON_SIZE);

        final GroupLayout.ParallelGroup components = layout.createParallelGroup()
                .addGroup(labels)
                .addGroup(button);

        return layout.createSequentialGroup()
                .addGap(HORIZONTAL_GAP_SIZE)
                .addGroup(components)
                .addGap(HORIZONTAL_GAP_SIZE);
    }

    /**
     * Returns vertical layout of components.
     *
     * @param layout layout
     * @return vertical layout of components
     */
    private GroupLayout.SequentialGroup createVerticalLayout(final GroupLayout layout) {
        return layout.createSequentialGroup()
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(nameLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE,
                        CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(versionLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE,
                        CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(authorLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE,
                        CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(rightsLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE,
                        CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_LONG_GAP_SIZE)
                .addComponent(okButton, CatalogSwingConstants.VERTICAL_BUTTON_SIZE, CatalogSwingConstants.VERTICAL_BUTTON_SIZE,
                        CatalogSwingConstants.VERTICAL_BUTTON_SIZE)
                .addGap(VERTICAL_GAP_SIZE);
    }

}

File: src/main/kotlin/cz/vhromada/catalog/gui/AboutDialog.kt
package cz.vhromada.catalog.gui

import cz.vhromada.catalog.gui.common.CatalogSwingConstants
import cz.vhromada.catalog.gui.common.Picture
import java.awt.Font
import javax.swing.GroupLayout
import javax.swing.JButton
import javax.swing.JDialog
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.SwingConstants
import javax.swing.WindowConstants

/**
 * A class represents dialog about.
 *
 * @author Vladimir Hromada
 */
class AboutDialog : JDialog(JFrame(), "About", true) {

    /**
     * Label for name
     */
    private val nameLabel = JLabel("Catalog", Picture.CATALOG.icon, SwingConstants.LEADING)

    /**
     * Label for version
     */
    private val versionLabel = JLabel("Version: 5.0.0")

    /**
     * Label for author
     */
    private val authorLabel = JLabel("Author: Vladimír Hromada")

    /**
     * Label for copyrights
     */
    private val rightsLabel = JLabel("All rights reserved.")

    /**
     * Button OK
     */
    private val okButton = JButton("OK", Picture.OK.icon)

    init {
        setIconImage(Picture.ABOUT.icon.image)
        defaultCloseOperation = WindowConstants.DISPOSE_ON_CLOSE
        isResizable = false

        val font = Font("Tahoma", Font.BOLD, FONT_SIZE)

        initLabels(font, nameLabel, versionLabel, authorLabel, rightsLabel)

        okButton.font = font
        okButton.addActionListener {
            isVisible = false
            dispose()
        }

        val layout = GroupLayout(contentPane)
        contentPane.layout = layout
        layout.setHorizontalGroup(createHorizontalLayout(layout))
        layout.setVerticalGroup(createVerticalLayout(layout))

        okButton.requestFocusInWindow()
        pack()
        setLocationRelativeTo(getRootPane())
    }

    /**
     * Returns horizontal layout of components.
     *
     * @param layout layout
     * @return horizontal layout of components
     */
    private fun createHorizontalLayout(layout: GroupLayout): GroupLayout.SequentialGroup {
        val labels = layout.createParallelGroup()
                .addComponent(nameLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE)
                .addComponent(versionLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE)
                .addComponent(authorLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE)
                .addComponent(rightsLabel, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE, HORIZONTAL_LABEL_SIZE)

        val button = layout.createSequentialGroup()
                .addGap(HORIZONTAL_BUTTON_GAP_SIZE)
                .addComponent(okButton, HORIZONTAL_BUTTON_SIZE, HORIZONTAL_BUTTON_SIZE, HORIZONTAL_BUTTON_SIZE)

        val components = layout.createParallelGroup()
                .addGroup(labels)
                .addGroup(button)

        return layout.createSequentialGroup()
                .addGap(HORIZONTAL_GAP_SIZE)
                .addGroup(components)
                .addGap(HORIZONTAL_GAP_SIZE)
    }

    /**
     * Returns vertical layout of components.
     *
     * @param layout layout
     * @return vertical layout of components
     */
    private fun createVerticalLayout(layout: GroupLayout): GroupLayout.SequentialGroup {
        return layout.createSequentialGroup()
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(nameLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(versionLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(authorLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_GAP_SIZE)
                .addComponent(rightsLabel, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE, CatalogSwingConstants.VERTICAL_COMPONENT_SIZE)
                .addGap(VERTICAL_LONG_GAP_SIZE)
                .addComponent(okButton, CatalogSwingConstants.VERTICAL_BUTTON_SIZE, CatalogSwingConstants.VERTICAL_BUTTON_SIZE, CatalogSwingConstants.VERTICAL_BUTTON_SIZE)
                .addGap(VERTICAL_GAP_SIZE)
    }


    /**
     * Initializes labels.
     *
     * @param font   font
     * @param labels labels
     */
    private fun initLabels(font: Font, vararg labels: JLabel) {
        for (label in labels) {
            label.isFocusable = false
            label.font = font
        }
    }

    companion object {

        /**
         * Horizontal label size
         */
        private const val HORIZONTAL_LABEL_SIZE = 200

        /**
         * Horizontal button size
         */
        private const val HORIZONTAL_BUTTON_SIZE = 76

        /**
         * Horizontal button gap size
         */
        private const val HORIZONTAL_BUTTON_GAP_SIZE = 62

        /**
         * Horizontal gap size
         */
        private const val HORIZONTAL_GAP_SIZE = 20

        /**
         * Vertical gap size
         */
        private const val VERTICAL_GAP_SIZE = 20

        /**
         * Vertical long gap size
         */
        private const val VERTICAL_LONG_GAP_SIZE = 40

        /**
         * Font size
         */
        private const val FONT_SIZE = 12

    }

}
